source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/stats/symbol_table.h"
+
+#include "common/common/assert.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Stats {
+
+class SymbolTableImpl : public SymbolTable {
+public:
+  std::vector<Symbol> encode(const std::string& name) override {
+    std::vector<Symbol> symbol_vec;
+    std::vector<std::string> name_vec = absl::StrSplit(name, '.');
+    symbol_vec.resize(name_vec.size());
+    std::transform(name_vec.begin(), name_vec.end(), symbol_vec.begin(),
+                   [this](std::string x) { return this->toSymbol(x); });
+    return symbol_vec;
+  }
+
+  std::string decode(const std::vector<Symbol>& symbol_vec) const override {
+    std::vector<std::string> name;
+    name.resize(symbol_vec.size());
+    std::transform(symbol_vec.begin(), symbol_vec.end(), name.begin(),
+                   [this](Symbol x) { return this->fromSymbol(x); });
+    return absl::StrJoin(name, ".");
+  }
+
+  bool free(const std::vector<Symbol>& symbol_vec) override {
+    bool successful_free = true;
+    for (const Symbol symbol : symbol_vec) {
+      successful_free = successful_free && freeSymbol(symbol);
+    }
+    return successful_free;
+  }
+
+  // For testing purposes only.
+  size_t size() const { return sizeof(SymbolTableImpl) + encode_map_.size() + decode_map_.size(); }
+
+private:
+  Symbol curr_counter_ = 0;
+
+  // Bimap implementation.
+  // The encode map stores both the symbol and the ref count of that symbol.
+  std::unordered_map<std::string, std::pair<Symbol, uint32_t>> encode_map_ = {};
```

Comment:
Do you have a plan for how to get this into shared memory with no heap allocation?  Isn't that the ultimate goal of this data structure?

Reply:
In anticipation of https://github.com/envoyproxy/envoy/pull/3710 which decouples the heap stats from shared-memory stats under the hood, we will only use this symbol table with the heap stat implementation.

Reply:
Gotcha.  That makes sense.  Thanks for the explanation.

Reply:
I'm thinking we need a better story here for how we will solve for shared memory, unless we want to make hot restart mutually exclusive with scalable stats (and scalable clusters). Istio, for example, sitll uses hot restart and wants to scale to very large cluster sizes. @mattklein123 @rshriram.

Reply:
That was my thinking -- that users of hot-restart didn't have stat-name scalability issues. But I didn't realize Istio might be a counter-example.  It's definitely way easier to solve this for non-hot-restart first in any case, and once #3710 is merged it is a separate problem.

Of course we don't have a problem making maps in shared memory -- BlockMemoryHashSet can be templated on  a pair and use only .first for hashing/comparing. But the annoying part is the symbol characters. We'd have to tune the number and size of the symbols. It's much easier in the non-hot-restart case which doesn't require tuning that.

Reply:
I would suggest that we discuss on the next community call.

FWIW I think it's probably OK to diverge the implementations between shared memory and the heap allocator, especially if we end up implementing https://github.com/envoyproxy/envoy/issues/3804 via the "light hot restart" proposal that @bplotnick is doing. For example, I can imagine a configuration in which the user uses the heap allocator to save memory, but still wants to use light hot restart, and just deals with the fact that there are stat inconsistencies during hot restart.